### PR TITLE
Updates for Compatibility with the new Cantera 3.0 Release

### PR DESCRIPTION
**Pull request by ecisneros8**
This PR updates the Pyrometheus code template and `chem_expr` module to ensure compatibility with the new Cantera release. 

The principal change in Cantera is that all reactions in a `ct.Solution` are of generic type `ct.Reaction`, no longer of any specific type (e.g., `ct.FalloffReaction`). Reaction types are now identified via the `ct.Reaction.reaction_type` interface. Pyrometheus has been updated accordingly to distinguish between elementary, falloff, and third-body reactions.

### CODE REVIEW

File: pyrometheus/chem_expr.py [hunk 1]
```diff
@@ -229,9 +229,23 @@ def third_body_efficiencies_expr(sol: ct.Solution, react: ct.Reaction, c):
...
         :class:`pymbolic.primitives.Expression`
     """
-    efficiencies = [react.efficiencies[sp] for sp in react.efficiencies]
-    indices_nondef = [sol.species_index(sp) for sp in react.efficiencies]
-    indices_default = [i for i in range(sol.n_species) if i not in indices_nondef]
+    if isinstance(react, ct.ThreeBodyReaction):
+        from warnings import warn
+        warn("The 'Reaction.efficiencies' interface is deprecated and "
+             "will be removed after Cantera 3. Access efficiencies via "
+             "'react.third_body.efficiencies' instead")
```
Comment (inducer) [inline pyrometheus/chem_expr.py:236]: I'm not sure we need to warn here. There's no thing actionable for the pyro user here.

File: pyrometheus/chem_expr.py [hunk 2]
```diff
@@ -243,10 +257,10 @@ def troe_falloff_expr(react: ct.Reaction, t):
     :returns: The Troe falloff center expression for reaction *react* in terms of the
         temperature *t* as a :class:`pymbolic.primitives.Expression`
     """
-    if not react.uses_legacy:
-        if react.rate.type == "Troe":
+    if "uses_legacy" not in dir(react) or not react.uses_legacy:
```
Comment (inducer) [inline pyrometheus/chem_expr.py:260]:
```suggestion
    if hasattr(react, "uses_legacy") or not react.uses_legacy:
```
(similar elsewhere)

File: pyrometheus/codegen/python.py [hunk 3]
```diff
@@ -476,6 +476,24 @@ def gen_thermochem_code(sol: ct.Solution) -> str:
     adhering to the :class:`~pyrometheus.thermochem_example.Thermochemistry`
     interface.
     """
+    if not all((isinstance(r, ct.Reaction) for r in sol.reactions())):
+        # Cantera version < 3.0
+        from warnings import warn
+        warn("Specific reaction types (e.g., ct.FalloffReaction) are "
+             "deprecated in Cantera 3, where all in a 'ct.Solution' "
+             "object are of generic 'ct.Reaction' type. "
+             "Identify reactions using 'ct.Reaction.reaction_type' insteady")
```
Comment (inducer) [inline pyrometheus/codegen/python.py:485]: As above.

File: pyrometheus/codegen/python.py [hunk 4]
```diff
@@ -427,7 +427,7 @@ def get_fwd_rate_coefficients(self, temperature, concentrations):
         ones = self._pyro_zeros_like(temperature) + 1.0
         k_fwd = [
         %for react in sol.reactions():
-        %if isinstance(react, ct.FalloffReaction):
+        %if react.equation in [r.equation for _, r in falloff_reactions]:
```
Comment (inducer) [inline pyrometheus/codegen/python.py:430]: This idiom is kind of inefficient. It seems like we could just test `react` instead?

File: test/test_codegen.py [hunk 5]
```diff
@@ -132,7 +136,22 @@ def test_get_rate_coefficients(mechname, usr_np):
...
+        # It seems like Cantera 3.0 has bumped the third-body efficiency
+        # factor from the rate coefficient to the rate of progress
+        for i, r in three_body_reactions:
+            eff = np.sum([c[sol.species_index(sp)]
+                          * r.third_body.efficiencies[sp]
+                          for sp in r.third_body.efficiencies])
+            eff += np.sum([c[sp] for sp in range(sol.n_species)
+                           if sol.species_name(sp) not in
+                           r.third_body.efficiencies])
+            k_ct[i] *= eff
```
Comment (inducer) [inline test/test_codegen.py:149]: Would this maintain compatibility with Cantera < 3?

**Review by inducer — COMMENTED** [review]
Thanks for working on this! My main question is whether we're hoping to maintain compatbility with cantera < 3. Some of the code is written as if that's the case, some other bits I'm not sure. If yes (i.e. maintain compatbility): why? The simpler thing to do might be to just make the jump (and update the dependency version in `setup.py`.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pyrometheus/codegen/python.py (hunk 4 above)
Comment (ecisneros8) [reply pyrometheus/codegen/python.py:430]: That was my initial approach but found that, for example, for the San Diego yaml in `mechs/test`:

`sol.reaction(9).equation == falloff_reactions[0].equation`

is true, so the 10th reaction is Falloff, and it is the first in the `falloff_reactions` list, but

`sol.Reaction(9) == falloff_reactions[0]`

is false, so testing `react in falloff_reactions` does not help here. I haven't been able to pin down what specific property makes them appear different upon comparison.

File: test/test_codegen.py (hunk 5 above)
Comment (ecisneros8) [reply test/test_codegen.py:149]: Thanks for noticing this. It will not. I will fix accordingly.

File: pyrometheus/chem_expr.py (hunk 1 above)
Comment (ecisneros8) [reply pyrometheus/chem_expr.py:236]: But we do have some users that modify the codegen. Indeed, @jonathan-macart has one such PR #73 !

**Comment by ecisneros8** [comment]
Thanks for reviewing this PR! That is a good point, and a simple way forward. 

I think that the strongest case for maintaining compatibility is to support our users that still use Cantera 2.6 in their toolchains. After all, Cantera 3 was release just 3 weeks ago or so. If we choose to support both versions, we will have to plan on when to drop < 3, correct?

**Comment by tulioricci** [comment]
Hi @ecisneros8, any issues towards merging this PR? Anything I could help with?

**Comment by ecisneros8** [comment]
Hi @tulioricci, thanks for pointing us back at this. We didn't merge this right away because the downstream Mirgecom test failed. My recollection is that it had to do with [building mpi4py](https://github.com/pyrometheus/pyrometheus/actions/runs/6426874713/job/17451685979#step:3:1303) during environment installing for the test. 

On the Pyro side everything is ready (but please confirm @inducer). If you can make the downstream test pass, it's a go. As a hint that may help, the Python version for tests is restricted to 3.11, because that's what Cantera supported when I created the PR.

**Comment by tulioricci** [comment]
Mirge-com tests will not pass while there is no mixture-averaged transport model inside pyrometheus.

**Comment by tulioricci** [comment]
> Hi @tulioricci, thanks for pointing us back at this. We didn't merge this right away because the downstream Mirgecom test failed. My recollection is that it had to do with [building mpi4py](https://github.com/pyrometheus/pyrometheus/actions/runs/6426874713/job/17451685979#step:3:1303) during environment installing for the test.

Mirge-com tests will not pass while there is no mixture-averaged transport model inside pyrometheus.

**Comment by matthiasdiener** [comment]
> Hi @tulioricci, thanks for pointing us back at this. We didn't merge this right away because the downstream Mirgecom test failed. My recollection is that it had to do with [building mpi4py](https://github.com/pyrometheus/pyrometheus/actions/runs/6426874713/job/17451685979#step:3:1303) during environment installing for the test.

#77 should fix the mpi4py-related error.

File: test/test_codegen.py [hunk 6]
```diff
@@ -110,7 +110,8 @@ def test_generate_mechfile(lang_module, mechname):
         print(code, file=mech_file)
 
 
-@pytest.mark.parametrize("mechname", ["uiuc.yaml", "sandiego.yaml", "uiuc.cti"])
+@pytest.mark.parametrize("mechname", ["uiuc.yaml", "sandiego.yaml",
+                                      "uiuc.yaml"])
```
Comment (tulioricci) [inline test/test_codegen.py:114]: Duplicated

File: test/test_codegen.py (hunk 5 above)
Comment (tulioricci) [inline test/test_codegen.py:149]: I have the impression that mirge-com hates *= and +=. Please, write this (and everywhere else) as `k_ct[i] = k_ct[i]*eff`

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -104,7 +104,7 @@ jobs:
                 curl -L -O https://tiker.net/ci-support-v0
                 . ./ci-support-v0
 
-                git clone "https://github.com/illinois-ceesd/$DOWNSTREAM_PROJECT.git"
+                git clone -b use_cantera30 "https://github.com/illinois-ceesd/$DOWNSTREAM_PROJECT.git"
```
Comment (tulioricci) [inline .github/workflows/ci.yml:107]: Let's not forget to undo this

**Review by tulioricci — APPROVED** [review]
LGTM. Just address my 2 comments and merge at your will.

I tested this in mirge-com as well and it seems to be working perfectly.

**Pull request merged**